common/network-types/src/udp.rs

```diff
@@ -1,152 +1,508 @@
-use std::io::Error;
+use futures::{pin_mut, ready, FutureExt, Sink, SinkExt};
+use std::io::ErrorKind;
 use std::pin::Pin;
+use std::sync::{Arc, OnceLock};
 use std::task::{Context, Poll};
-use tokio::io::ReadBuf;
+use tokio::io::{AsyncRead, ReadBuf};
 use tokio::net::UdpSocket;
-use tracing::{trace, warn};
+use tracing::{debug, error, trace, warn};
+
+type BoxIoSink<T> = Box<dyn Sink<T, Error = std::io::Error> + Send + Unpin>;
 
 /// Mimics TCP-like stream functionality on a UDP socket by restricting it to a single
 /// counterparty and implements [`tokio::io::AsyncRead`] and [`tokio::io::AsyncWrite`].
+/// The instance is always constructed using a [`UdpStreamBuilder`].
 ///
 /// To set a counterparty, one of the following must happen:
-/// 1) setting it via [`with_counterparty`](ConnectedUdpStream::with_counterparty) after binding
+/// 1) setting it during build via [`UdpStreamBuilder::with_counterparty`]
 /// 2) receiving some data from the other side.
 ///
 /// Whatever of the above happens first, sets the counterparty.
 /// Once the counterparty is set, all data sent and received will be sent or filtered by this
 /// counterparty address.
 ///
 /// If data from another party is received, an error is raised, unless the object has been constructed
-/// with [`with_foreign_data_discarding`](ConnectedUdpStream::with_foreign_data_discarding) setting.
-#[derive(Debug)]
+/// with [`ForeignDataMode::Discard`] or [`ForeignDataMode::Accept`] setting.
+///
+/// This object is also capable of parallel processing on a UDP socket.
+/// If [parallelism](UdpStreamBuilder::with_parallelism) is set, the instance will create
+/// multiple sockets with `SO_REUSEADDR` and spin parallel tasks that will coordinate data and
+/// transmission retrieval using these sockets. This is driven by RX/TX MPMC queues, which are
+/// per-default unbounded (see [queue size](UdpStreamBuilder::with_queue_size) for details).
 pub struct ConnectedUdpStream {
-    sock: UdpSocket,
-    counterparty: Option<std::net::SocketAddr>,
-    closed: bool,
-    foreign_data_mode: ForeignDataMode,
+    socket_handles: Vec<(tokio::task::JoinHandle<()>, tokio::task::JoinHandle<()>)>,
```

Comment:
`tokio` only implementation?

Reply:
Currently yes, but it can be also made `async-std` friendly as well, with not much effort.